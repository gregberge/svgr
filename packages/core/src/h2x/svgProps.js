import { JSXAttribute } from 'h2x-plugin-jsx'

const compareAttrsName = attr => otherAttr => attr.name === otherAttr.name
const compareAttrsValue = attr => otherAttr => attr.value === otherAttr.value

const compareAttrs = attr => otherAttr =>
  compareAttrsName(attr)(otherAttr) && compareAttrsValue(attr)(otherAttr)

const areAttrsAlreadyInjected = (node, attributes = {}) => {
  const nodeAttrs = node.attributes

  return Object.keys(attributes).reduce((accumulation, key) => {
    if (nodeAttrs.some(compareAttrs({ name: key, value: attributes[key] })))
      return accumulation
    return false
  }, true)
}

const svgProps = (props = {}) => () => {
  const interpolated = new Set()
  const keys = Object.keys(props)
  const attributes = keys.reduce((acc, prop) => {
    const value = props[prop]
    if (
      typeof value === 'string' &&
      value.startsWith('{') &&
      value.endsWith('}')
    ) {
      acc[prop] = value.slice(1, -1)
      interpolated.add(prop)
    } else {
      acc[prop] = value
    }
    return acc
  }, {})

  return {
    visitor: {
      JSXElement: {
        enter(path) {
          if (path.node.name !== 'svg') return
          if (areAttrsAlreadyInjected(path.node, attributes)) return

          const parseAttributes = keys.reduce((accumulation, key) => {
            const prop = new JSXAttribute()
            prop.name = key
            prop.value = attributes[key]
            prop.literal = interpolated.has(key)
            return [...accumulation, prop]
          }, [])

          const mergeAttributes = path.node.attributes.reduce(
            (accumulation, value) => {
              if (accumulation.some(compareAttrsName(value)))
                return accumulation
              return [...accumulation, value]
            },
            parseAttributes,
          )

          path.node.attributes = mergeAttributes
          path.replace(path.node)
        },
      },
    },
  }
}

export default svgProps
