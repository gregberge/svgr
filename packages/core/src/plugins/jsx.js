import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
