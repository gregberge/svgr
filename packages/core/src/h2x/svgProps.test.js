import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import svgProps from './svgProps'

function transformSvgProps(code, props) {
  return transform(code, { plugins: [jsx, svgProps(props)] })
}

describe('svgProps', () => {
  it('should add one prop to svg', () => {
    expect(
      transformSvgProps(
        ` <svg>
            <g stroke="#063855" stroke-width="2">
              <path d="M51,37 L37,51"></path>
            </g>
          </svg>
        `,
        { focusable: false },
      ),
    ).toMatchInlineSnapshot(`
"<svg focusable=\\"false\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`)
  })

  it('should add multiple props to svg', () => {
    expect(
      transformSvgProps(
        ` <svg>
            <g stroke="#063855" stroke-width="2">
              <path d="M51,37 L37,51"></path>
            </g>
          </svg>
        `,
        {
          focusable: false,
          hidden: 'hidden',
        },
      ),
    ).toMatchInlineSnapshot(`
"<svg focusable=\\"false\\" hidden=\\"hidden\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`)
  })

  it('should update old prop to the new prop', () => {
    expect(
      transformSvgProps(
        ` <svg focusable="false">
            <g stroke="#063855" stroke-width="2">
              <path d="M51,37 L37,51"></path>
            </g>
          </svg>
        `,
        { focusable: true },
      ),
    ).toMatchInlineSnapshot(`
"<svg focusable=\\"true\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`)
  })

  it('should add a new prop', () => {
    expect(
      transformSvgProps(
        ` <svg focusable="true" color="#fff">
            <g stroke="#063855" stroke-width="2">
              <path d="M51,37 L37,51"></path>
            </g>
          </svg>
        `,
        { focusable: true, hidden: true },
      ),
    ).toMatchInlineSnapshot(`
"<svg focusable=\\"true\\" hidden=\\"true\\" color=\\"#fff\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`)
  })

  it('should interpolate prop with curly brackets', () => {
    expect(
      transformSvgProps(
        ` <svg focusable="true" color="#fff">
            <g stroke="#063855" stroke-width="2">
              <path d="M51,37 L37,51"></path>
            </g>
          </svg>
        `,
        {
          // literal
          focusable: `{true}`,
          // variables
          hidden: `{hidden}`,
          // expression
          fill: `{fill == null ? 'currentColor' : fill}`,
          // string
          id: 'foo',
        },
      ),
    ).toMatchInlineSnapshot(`
"<svg focusable={true} hidden={hidden} fill={fill == null ? 'currentColor' : fill} id=\\"foo\\" color=\\"#fff\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" />
  </g>
</svg>
"
`)
  })
})
