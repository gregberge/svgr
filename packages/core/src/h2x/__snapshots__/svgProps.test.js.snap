// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgProps should add a new prop 1`] = `
"<svg focusable={true} hidden={true} color=\\"#fff\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgProps should add multiple props to svg 1`] = `
"<svg focusable={false} hidden={hidden}>
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgProps should add one prop to svg 1`] = `
"<svg focusable={false}>
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgProps should update old prop to the new prop 1`] = `
"<svg focusable={true}>
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;
